// ng2-date-picker/esm5/date-picker/date-picker.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { DatePickerDirectiveService } from './date-picker-directive.service';
import { DatePickerComponent } from './date-picker.component';
import { ComponentFactoryResolver, Directive, ElementRef, EventEmitter, HostListener, Input, Optional, Output, ViewContainerRef } from '@angular/core';
import { NgControl } from '@angular/forms';
import { UtilsService } from '../common/services/utils/utils.service';
var DatePickerDirective = /** @class */ (function () {
    function DatePickerDirective(viewContainerRef, elemRef, componentFactoryResolver, service, formControl, utilsService) {
        this.viewContainerRef = viewContainerRef;
        this.elemRef = elemRef;
        this.componentFactoryResolver = componentFactoryResolver;
        this.service = service;
        this.formControl = formControl;
        this.utilsService = utilsService;
        this._mode = 'day';
        this.open = new EventEmitter();
        this.close = new EventEmitter();
        this.onChange = new EventEmitter();
        this.onGoToCurrent = new EventEmitter();
        this.onLeftNav = new EventEmitter();
        this.onRightNav = new EventEmitter();
        this.onSelect = new EventEmitter();
    }
    Object.defineProperty(DatePickerDirective.prototype, "config", {
        get: /**
         * @return {?}
         */
        function () {
            return this._config;
        },
        set: /**
         * @param {?} config
         * @return {?}
         */
        function (config) {
            this._config = this.service.getConfig(config, this.viewContainerRef.element, this.attachTo);
            this.updateDatepickerConfig();
            this.markForCheck();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DatePickerDirective.prototype, "attachTo", {
        get: /**
         * @return {?}
         */
        function () {
            return this._attachTo;
        },
        set: /**
         * @param {?} attachTo
         * @return {?}
         */
        function (attachTo) {
            this._attachTo = attachTo;
            this._config = this.service.getConfig(this.config, this.viewContainerRef.element, this.attachTo);
            this.updateDatepickerConfig();
            this.markForCheck();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DatePickerDirective.prototype, "theme", {
        get: /**
         * @return {?}
         */
        function () {
            return this._theme;
        },
        set: /**
         * @param {?} theme
         * @return {?}
         */
        function (theme) {
            this._theme = theme;
            if (this.datePicker) {
                this.datePicker.theme = theme;
            }
            this.markForCheck();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DatePickerDirective.prototype, "mode", {
        get: /**
         * @return {?}
         */
        function () {
            return this._mode;
        },
        set: /**
         * @param {?} mode
         * @return {?}
         */
        function (mode) {
            this._mode = mode;
            if (this.datePicker) {
                this.datePicker.mode = mode;
            }
            this.markForCheck();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DatePickerDirective.prototype, "minDate", {
        get: /**
         * @return {?}
         */
        function () {
            return this._minDate;
        },
        set: /**
         * @param {?} minDate
         * @return {?}
         */
        function (minDate) {
            this._minDate = minDate;
            if (this.datePicker) {
                this.datePicker.minDate = minDate;
                this.datePicker.ngOnInit();
            }
            this.markForCheck();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DatePickerDirective.prototype, "maxDate", {
        get: /**
         * @return {?}
         */
        function () {
            return this._maxDate;
        },
        set: /**
         * @param {?} maxDate
         * @return {?}
         */
        function (maxDate) {
            this._maxDate = maxDate;
            if (this.datePicker) {
                this.datePicker.maxDate = maxDate;
                this.datePicker.ngOnInit();
            }
            this.markForCheck();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DatePickerDirective.prototype, "minTime", {
        get: /**
         * @return {?}
         */
        function () {
            return this._minTime;
        },
        set: /**
         * @param {?} minTime
         * @return {?}
         */
        function (minTime) {
            this._minTime = minTime;
            if (this.datePicker) {
                this.datePicker.minTime = minTime;
                this.datePicker.ngOnInit();
            }
            this.markForCheck();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DatePickerDirective.prototype, "maxTime", {
        get: /**
         * @return {?}
         */
        function () {
            return this._maxTime;
        },
        set: /**
         * @param {?} maxTime
         * @return {?}
         */
        function (maxTime) {
            this._maxTime = maxTime;
            if (this.datePicker) {
                this.datePicker.maxTime = maxTime;
                this.datePicker.ngOnInit();
            }
            this.markForCheck();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DatePickerDirective.prototype, "displayDate", {
        get: /**
         * @return {?}
         */
        function () {
            return this._displayDate;
        },
        set: /**
         * @param {?} displayDate
         * @return {?}
         */
        function (displayDate) {
            this._displayDate = displayDate;
            this.updateDatepickerConfig();
            this.markForCheck();
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    DatePickerDirective.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.datePicker = this.createDatePicker();
        this.api = this.datePicker.api;
        this.updateDatepickerConfig();
        this.attachModelToDatePicker();
        this.datePicker.theme = this.theme;
    };
    /**
     * @return {?}
     */
    DatePickerDirective.prototype.createDatePicker = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var factory = this.componentFactoryResolver.resolveComponentFactory(DatePickerComponent);
        return this.viewContainerRef.createComponent(factory).instance;
    };
    /**
     * @return {?}
     */
    DatePickerDirective.prototype.attachModelToDatePicker = /**
     * @return {?}
     */
    function () {
        var _this = this;
        if (!this.formControl) {
            return;
        }
        this.datePicker.onViewDateChange(this.formControl.value);
        this.formControl.valueChanges.subscribe(function (value) {
            if (value !== _this.datePicker.inputElementValue) {
                /** @type {?} */
                var strVal = _this.utilsService.convertToString(value, _this.datePicker.componentConfig.format);
                _this.datePicker.onViewDateChange(strVal);
            }
        });
        /** @type {?} */
        var setup = true;
        this.datePicker.registerOnChange(function (value, changedByInput) {
            if (value) {
                /** @type {?} */
                var isMultiselectEmpty = setup && Array.isArray(value) && !value.length;
                if (!isMultiselectEmpty && !changedByInput) {
                    _this.formControl.control.setValue(_this.datePicker.inputElementValue);
                }
            }
            /** @type {?} */
            var errors = _this.datePicker.validateFn(value);
            if (!setup) {
                _this.formControl.control.markAsDirty({
                    onlySelf: true
                });
            }
            else {
                setup = false;
            }
            if (errors) {
                if (errors.hasOwnProperty('format')) {
                    var given = errors['format'].given;
                    _this.datePicker.inputElementValue = given;
                    if (!changedByInput) {
                        _this.formControl.control.setValue(given);
                    }
                }
                _this.formControl.control.setErrors(errors);
            }
        });
    };
    /**
     * @return {?}
     */
    DatePickerDirective.prototype.onClick = /**
     * @return {?}
     */
    function () {
        this.datePicker.onClick();
    };
    /**
     * @return {?}
     */
    DatePickerDirective.prototype.onFocus = /**
     * @return {?}
     */
    function () {
        this.datePicker.inputFocused();
    };
    /**
     * @private
     * @return {?}
     */
    DatePickerDirective.prototype.updateDatepickerConfig = /**
     * @private
     * @return {?}
     */
    function () {
        if (this.datePicker) {
            this.datePicker.minDate = this.minDate;
            this.datePicker.maxDate = this.maxDate;
            this.datePicker.minTime = this.minTime;
            this.datePicker.maxTime = this.maxTime;
            this.datePicker.mode = this.mode || 'day';
            this.datePicker.displayDate = this.displayDate;
            this.datePicker.config = this.config;
            this.datePicker.open = this.open;
            this.datePicker.close = this.close;
            this.datePicker.onChange = this.onChange;
            this.datePicker.onGoToCurrent = this.onGoToCurrent;
            this.datePicker.onLeftNav = this.onLeftNav;
            this.datePicker.onRightNav = this.onRightNav;
            this.datePicker.onSelect = this.onSelect;
            this.datePicker.init();
            if (this.datePicker.componentConfig.disableKeypress) {
                this.elemRef.nativeElement.setAttribute('readonly', true);
            }
            else {
                this.elemRef.nativeElement.removeAttribute('readonly');
            }
        }
    };
    /**
     * @return {?}
     */
    DatePickerDirective.prototype.markForCheck = /**
     * @return {?}
     */
    function () {
        if (this.datePicker) {
            this.datePicker.cd.markForCheck();
        }
    };
    DatePickerDirective.decorators = [
        { type: Directive, args: [{
                    exportAs: 'dpDayPicker',
                    providers: [DatePickerDirectiveService],
                    selector: '[dpDayPicker]'
                },] }
    ];
    /** @nocollapse */
    DatePickerDirective.ctorParameters = function () { return [
        { type: ViewContainerRef },
        { type: ElementRef },
        { type: ComponentFactoryResolver },
        { type: DatePickerDirectiveService },
        { type: NgControl, decorators: [{ type: Optional }] },
        { type: UtilsService }
    ]; };
    DatePickerDirective.propDecorators = {
        config: [{ type: Input, args: ['dpDayPicker',] }],
        attachTo: [{ type: Input }],
        theme: [{ type: Input }],
        mode: [{ type: Input }],
        minDate: [{ type: Input }],
        maxDate: [{ type: Input }],
        minTime: [{ type: Input }],
        maxTime: [{ type: Input }],
        displayDate: [{ type: Input }],
        open: [{ type: Output }],
        close: [{ type: Output }],
        onChange: [{ type: Output }],
        onGoToCurrent: [{ type: Output }],
        onLeftNav: [{ type: Output }],
        onRightNav: [{ type: Output }],
        onSelect: [{ type: Output }],
        onClick: [{ type: HostListener, args: ['click',] }],
        onFocus: [{ type: HostListener, args: ['focus',] }]
    };
    return DatePickerDirective;
}());
export { DatePickerDirective };
if (false) {
    /**
     * @type {?}
     * @private
     */
    DatePickerDirective.prototype._config;
    /**
     * @type {?}
     * @private
     */
    DatePickerDirective.prototype._attachTo;
    /**
     * @type {?}
     * @private
     */
    DatePickerDirective.prototype._theme;
    /**
     * @type {?}
     * @private
     */
    DatePickerDirective.prototype._mode;
    /**
     * @type {?}
     * @private
     */
    DatePickerDirective.prototype._minDate;
    /**
     * @type {?}
     * @private
     */
    DatePickerDirective.prototype._maxDate;
    /**
     * @type {?}
     * @private
     */
    DatePickerDirective.prototype._minTime;
    /**
     * @type {?}
     * @private
     */
    DatePickerDirective.prototype._maxTime;
    /**
     * @type {?}
     * @private
     */
    DatePickerDirective.prototype._displayDate;
    /** @type {?} */
    DatePickerDirective.prototype.open;
    /** @type {?} */
    DatePickerDirective.prototype.close;
    /** @type {?} */
    DatePickerDirective.prototype.onChange;
    /** @type {?} */
    DatePickerDirective.prototype.onGoToCurrent;
    /** @type {?} */
    DatePickerDirective.prototype.onLeftNav;
    /** @type {?} */
    DatePickerDirective.prototype.onRightNav;
    /** @type {?} */
    DatePickerDirective.prototype.onSelect;
    /** @type {?} */
    DatePickerDirective.prototype.datePicker;
    /** @type {?} */
    DatePickerDirective.prototype.api;
    /** @type {?} */
    DatePickerDirective.prototype.viewContainerRef;
    /** @type {?} */
    DatePickerDirective.prototype.elemRef;
    /** @type {?} */
    DatePickerDirective.prototype.componentFactoryResolver;
    /** @type {?} */
    DatePickerDirective.prototype.service;
    /** @type {?} */
    DatePickerDirective.prototype.formControl;
    /** @type {?} */
    DatePickerDirective.prototype.utilsService;
}
